File: release/ci-deploy-bintray.rb
#!/usr/bin/env ruby
# uploads .deb and .rpm files to bintray

require_relative 'common'

module Itch
  def Itch.ci_deploy_bintray
    say "Should deploy to Bintray!"
    ✓ gem_dep 'dpl', 'dpl'

    FileUtils.mkdir_p "build"

    release_date = Time.now.strftime('%Y-%m-%d')
    %w(rpm deb).each do |repo|
      %w(386 amd64).each do |arch|
        say "Uploading #{arch} to #{repo} repo..."
        deb_arch = to_deb_arch arch
        rpm_arch = case arch
                   when "386" then "i386"
                   when "amd64" then "x86_64"
                   else raise "Unsupported arch #{arch}"
                   end

        publish = (channel_name != 'stable')
        say "(and publishing)" if publish

        conf = File.read("release/templates/bintray.#{repo}.json.in")
        conf = conf.gsub "{{CI_APPNAME}}", app_name
        conf = conf.gsub "{{CI_VERSION}}", build_version
        conf = conf.gsub "{{CI_RELEASE_DATE}}", release_date
        conf = conf.gsub "{{CI_PUBLISH}}", publish.to_s
        conf = conf.gsub "{{DEB_ARCH}}", deb_arch
        conf = conf.gsub "{{RPM_ARCH}}", rpm_arch
        File.write("build/bintray.json", conf)

        ✓ qsh %Q{dpl --provider=bintray --file=build/bintray.json --user=fasterthanlime --key="#{ENV['BINTRAY_TOKEN']}"}
      end # 386, amd64
    end # deb, rpm
  end
end

Itch.ci_deploy_bintray

File: release/ci-deploy-bintray.js
#!/usr/bin/env node

// uploads .deb and .rpm files to bintray

const $ = require("./common");

async function deployBintray() {
  if ($.channelName() === "canary") {
    $.say("Skipping bintray deploy for canary versions");
    process.exit(0);
  }

  $.say("Deploying to Bintray!");
  await $.gemDep("dpl", "dpl");

  $(await $.sh("mkdir -p build"));

  var date = $.buildTime();
  const releaseDate = `${date.getUTCFullYear()}-${$.leftPad(
    "" + date.getUTCMonth(),
    2,
    "0"
  )}-${$.leftPad("" + date.getUTCDate(), 2, "0")}`;
  $.say(`Release date: ${releaseDate}`);

  const repos = ["rpm", "deb"];
  const arches = ["386", "amd64"];
  for (const repo of repos) {
    for (const arch of arches) {
      $.say(`Uploading ${arch} to ${repo} repo...`);

      const debArch = $.toDebArch(arch);
      const rpmArch = $.toRpmArch(arch);

      // auto-publish releases, it's too easy to forget flipping the switch otherwise
      const publish = true;

      let conf = await $.readFile(`release/templates/bintray.${repo}.json.in`);
      conf = conf.replace(/{{CI_APPNAME}}/g, $.appName());
      conf = conf.replace(/{{CI_VERSION}}/g, $.buildVersion());
      conf = conf.replace(/{{CI_RELEASE_DATE}}/g, releaseDate);
      conf = conf.replace(/{{CI_PUBLISH}}/g, "" + publish);
      conf = conf.replace(/{{DEB_ARCH}}/g, debArch);
      conf = conf.replace(/{{RPM_ARCH}}/g, rpmArch);
      await $.writeFile("build/bintray.json", conf);

      $(
        await $.qsh(
          `dpl --provider=bintray --file=build/bintray.json --user=fasterthanlime --key="${process
            .env.BINTRAY_TOKEN}"`
        )
      );
    }
  }
}

deployBintray();
